Assignment 3 Sampling
Name - Siddharth Chaudhary
Roll no. - 102003707
Group - 3CO28

In [1]:
import pandas as pd

In [2]:
URL = "https://github.com/AnjulaMehto/Sampling_Assignment/blob/main/Creditcard_data.csv"

About Data
The dataset is highly unbalanced, the positive class (frauds) account for 1.172% of all transactions. It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, the original features and more background information about the data could not be provided. Features V1, V2, ... V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

In [3]:
data = pd.read_csv("C:\\Users\\SIDDHARTH CHAUDHARY\\Documents\\Sampling_Assignment-main\\Creditcard_data.csv")

In [4]:
data

Time        V1        V2        V3        V4        V5        V6  \
0       0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1       0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2       1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3       1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4       2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
..    ...       ...       ...       ...       ...       ...       ...   
767   575 -0.572263  0.731748  1.541254  0.150506  1.108974  0.372152   
768   579 -1.296845 -0.511605  2.404726 -0.310762 -0.319551 -0.542842   
769   579  1.214170  0.210481  0.484651  0.479768 -0.261955 -0.527039   
770   580  1.267030 -0.071114  0.037680  0.512683  0.242392  0.705212   
771   581  1.153758  0.132273  0.382969  1.405063 -0.224287 -0.197295   

           V7        V8        V9  ...       V21       V22       V23  \
0    0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1   -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2    0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3    0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4    0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
..        ...       ...       ...  ...       ...       ...       ...   
767  1.084879 -0.146329 -0.274447  ... -0.143508 -0.107582 -0.418263   
768 -0.173310  0.260423 -1.202688  ... -0.071270 -0.161175  0.088496   
769  0.021782 -0.106888 -0.037631  ... -0.224292 -0.594609  0.159877   
770 -0.226582  0.109483  0.657565  ... -0.164468 -0.177225 -0.222918   
771  0.020653  0.029260  0.412254  ... -0.107809 -0.125231 -0.057041   

          V24       V25       V26       V27       V28  Amount  Class  
0    0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1   -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
2   -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3   -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4    0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
..        ...       ...       ...       ...       ...     ...    ...  
767 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   26.72      0  
768  0.285390  0.281069 -0.370130  0.043410  0.092318   80.00      0  
769  0.091873  0.140964  0.227406 -0.017389  0.016030    5.98      0  
770 -1.245505  0.678360  0.525059  0.002920 -0.003333   12.36      0  
771  0.073082  0.633977 -0.310685  0.033590  0.015250   13.79      0  

[772 rows x 31 columns]

In [6]:
import imblearn
print(imblearn.__version__)

0.10.1


In [11]:
import warnings
warnings.filterwarnings('ignore')
from IPython.display import Image

#Libraries used
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#sns.set_style("darkgrid")

import plotly_express as px
import plotly.graph_objects as go
import plotly.figure_factory as ff
from plotly.subplots import make_subplots

# Classifier Libraries
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import collections

from imblearn.under_sampling import RandomUnderSampler
from sklearn.preprocessing import MinMaxScaler, RobustScaler
from sklearn.model_selection import StratifiedKFold, train_test_split, cross_val_score, GridSearchCV, RandomizedSearchCV
from sklearn.model_selection import learning_curve
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from imblearn.pipeline import make_pipeline as imbalanced_make_pipeline
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import NearMiss
from scipy.stats import norm
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import roc_auc_score, roc_curve
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

import keras
from keras.models import Sequential
from keras.layers import Activation
from keras.layers.core import Dense
from keras.optimizers import Adam
from keras.metrics import categorical_crossentropy

%matplotlib inline

In [12]:
df = data
df.info

<bound method DataFrame.info of      Time        V1        V2        V3        V4        V5        V6  \
0       0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1       0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2       1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3       1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4       2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
..    ...       ...       ...       ...       ...       ...       ...   
767   575 -0.572263  0.731748  1.541254  0.150506  1.108974  0.372152   
768   579 -1.296845 -0.511605  2.404726 -0.310762 -0.319551 -0.542842   
769   579  1.214170  0.210481  0.484651  0.479768 -0.261955 -0.527039   
770   580  1.267030 -0.071114  0.037680  0.512683  0.242392  0.705212   
771   581  1.153758  0.132273  0.382969  1.405063 -0.224287 -0.197295   

           V7        V8        V9  ...       V21       V22       V23  \
0    0.239599  0.09

In [13]:
# The classes are heavily skewed we need to solve this issue later.
print('No Frauds', round(df['Class'].value_counts()[0]/len(df) * 100,2), '% of the dataset')
print('Frauds', round(df['Class'].value_counts()[1]/len(df) * 100,2), '% of the dataset')

No Frauds 98.83 % of the dataset
Frauds 1.17 % of the dataset


98.83% -> No Frauds
1.17% -> Frauds

Since, the dataset contains no null or missing values and all the features seem to be scaled properly, there seems to be no need of any preprocessing. Only the features Time & Amount are required to be scaled.

In [20]:
rob_scaler = RobustScaler()
df['scaled_time'] = rob_scaler.fit_transform(df['Time'].values.reshape(-1,1))
df['scaled_amount'] = rob_scaler.fit_transform(df['Amount'].values.reshape(-1,1))
df.drop(['Time','Amount'], axis=1, inplace=True)

In [21]:
scaled_amount = df['scaled_amount']
scaled_time = df['scaled_time']

#scaled_amount and scaled_time are added to the starting of the dataframe
df.drop(['scaled_amount', 'scaled_time'], axis=1, inplace=True)
df.insert(0, 'scaled_amount', scaled_amount)
df.insert(1, 'scaled_time', scaled_time)

Dataset After Scaling Time and Amount

In [22]:
df

scaled_amount  scaled_time        V1        V2        V3        V4  \
0         2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1        -0.282095    -0.923077  1.191857  0.266151  0.166480  0.448154   
2         7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
3         2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
4         1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   
..             ...          ...       ...       ...       ...       ...   
767       0.202967     0.959083 -0.572263  0.731748  1.541254  0.150506   
768       1.278462     0.972177 -1.296845 -0.511605  2.404726 -0.310762   
769      -0.215684     0.972177  1.214170  0.210481  0.484651  0.479768   
770      -0.086899     0.975450  1.267030 -0.071114  0.037680  0.512683   
771      -0.058034     0.978723  1.153758  0.132273  0.382969  1.405063   

           V5        V6        V7        V8  ...       V20       V21  \
0   -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307   
1    0.060018 -0.082361 -0.078803  0.085102  ... -0.069083 -0.225775   
2   -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998   
3   -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300   
4   -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431   
..        ...       ...       ...       ...  ...       ...       ...   
767  1.108974  0.372152  1.084879 -0.146329  ...  0.056544 -0.143508   
768 -0.319551 -0.542842 -0.173310  0.260423  ... -0.349668 -0.071270   
769 -0.261955 -0.527039  0.021782 -0.106888  ... -0.071540 -0.224292   
770  0.242392  0.705212 -0.226582  0.109483  ... -0.066990 -0.164468   
771 -0.224287 -0.197295  0.020653  0.029260  ... -0.230877 -0.107809   

          V22       V23       V24       V25       V26       V27       V28  \
0    0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1   -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2    0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3    0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4    0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
..        ...       ...       ...       ...       ...       ...       ...   
767 -0.107582 -0.418263 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   
768 -0.161175  0.088496  0.285390  0.281069 -0.370130  0.043410  0.092318   
769 -0.594609  0.159877  0.091873  0.140964  0.227406 -0.017389  0.016030   
770 -0.177225 -0.222918 -1.245505  0.678360  0.525059  0.002920 -0.003333   
771 -0.125231 -0.057041  0.073082  0.633977 -0.310685  0.033590  0.015250   

     Class  
0        0  
1        1  
2        0  
3        0  
4        0  
..     ...  
767      0  
768      0  
769      0  
770      0  
771      0  

[772 rows x 31 columns]

Balancing the data
Now I will perform EDA on the dataset to study the various features of it. Since, it is an imbalanced dataset, I would first undersample the majority class(i.e. Non-Fraud in this case) so that I can get an equal distribution of Fraud and Non-Fraud cases. For undersampling I have used Random under sampling before crossvalidation.


Random Under Sampling

In [23]:
#Random Under Sampling
X = df.drop('Class', axis=1)
y = df['Class']
rus = RandomUnderSampler(random_state=42)
X_rs, y_rs = rus.fit_resample(X,y)

In [24]:
#Random Under-sampled Dataframe
df_rs = pd.DataFrame(np.hstack((X_rs,y_rs[:, None])), columns=df.columns)
df_rs.Class = df_rs.Class.astype(int)
df_rs.head()

scaled_amount  scaled_time        V1        V2        V3        V4  \
0      -0.154118    -0.055646 -0.491003  0.906953  1.645423 -0.083531   
1      -0.296427    -0.320786 -0.143256  0.743649  1.534072  1.062170   
2       0.038252     0.929624  1.166360  0.005061  0.497768  0.798920   
3      -0.158761    -0.504092  1.239495 -0.182609  0.155058 -0.928892   
4      -0.276241    -0.117840 -1.069200  1.239963  0.545157  1.005354   

         V5        V6        V7        V8  ...       V20       V21       V22  \
0 -0.195560 -0.710165  0.559119  0.116340  ...  0.062757 -0.168067 -0.517387   
1  0.208187 -0.838623  0.524151 -0.294661  ...  0.271423 -0.205014 -0.460893   
2 -0.365524 -0.233421 -0.074210 -0.008325  ... -0.118735 -0.195728 -0.365798   
3 -0.746227 -1.235608 -0.061695 -0.125223  ... -0.160406  0.146077  0.481119   
4 -0.025696 -0.910673  0.422442  0.049283  ... -0.134162  0.146555  0.602990   

        V23       V24       V25       V26       V27       V28  Class  
0  0.018650  0.491652 -0.277795  0.043841  0.253372  0.111749      0  
1  0.047407  0.339243 -0.779439  0.234456 -0.024125 -0.049898      0  
2  0.030729  0.123133  0.381749  0.296735 -0.007175  0.011905      0  
3 -0.140019  0.538261  0.710720 -0.621382  0.036867  0.010963      0  
4  0.132656  0.427113 -0.084030 -0.417194 -0.897885 -0.462042      0  

[5 rows x 31 columns]

In [ ]:
#Frauds vs. Non-Frauds for the new dataframe
group = df_rs.groupby('Class')['scaled_time'].count()
fig = go.Figure()
fig.add_trace(go.Bar(name='counts',x=group.index, y=group.values, marker_line_color='black', marker_line_width=2,
                     text=group.values,textposition='outside'))
fig.update_yaxes(title_text='Number of Cases')
fig.update_xaxes(title_text='Case',nticks=3)
fig.update_layout(template='seaborn',hovermode='closest',title='Number of Fraud vs Non-Fraud Cases',
                 width=700,height=400,xaxis=dict(mirror=True,linecolor='black',linewidth=2),
                 yaxis=dict(mirror=True,linecolor='black',linewidth=2),margin=dict(t=50,b=0,l=70,r=0))
fig.show()

In [26]:
print('Percentage of No Frauds: {}%'.format(round(df_rs.Class.value_counts()[0]/len(df_rs) * 100.0,2)))
print('Percentage of Frauds: {}%'.format(round(df_rs.Class.value_counts()[1]/len(df_rs) * 100.0,2)))

Percentage of No Frauds: 50.0%
Percentage of Frauds: 50.0%



Smote Oversampling

In [27]:
df2=data
df2

scaled_amount  scaled_time        V1        V2        V3        V4  \
0         2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1        -0.282095    -0.923077  1.191857  0.266151  0.166480  0.448154   
2         7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
3         2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
4         1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   
..             ...          ...       ...       ...       ...       ...   
767       0.202967     0.959083 -0.572263  0.731748  1.541254  0.150506   
768       1.278462     0.972177 -1.296845 -0.511605  2.404726 -0.310762   
769      -0.215684     0.972177  1.214170  0.210481  0.484651  0.479768   
770      -0.086899     0.975450  1.267030 -0.071114  0.037680  0.512683   
771      -0.058034     0.978723  1.153758  0.132273  0.382969  1.405063   

           V5        V6        V7        V8  ...       V20       V21  \
0   -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307   
1    0.060018 -0.082361 -0.078803  0.085102  ... -0.069083 -0.225775   
2   -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998   
3   -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300   
4   -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431   
..        ...       ...       ...       ...  ...       ...       ...   
767  1.108974  0.372152  1.084879 -0.146329  ...  0.056544 -0.143508   
768 -0.319551 -0.542842 -0.173310  0.260423  ... -0.349668 -0.071270   
769 -0.261955 -0.527039  0.021782 -0.106888  ... -0.071540 -0.224292   
770  0.242392  0.705212 -0.226582  0.109483  ... -0.066990 -0.164468   
771 -0.224287 -0.197295  0.020653  0.029260  ... -0.230877 -0.107809   

          V22       V23       V24       V25       V26       V27       V28  \
0    0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1   -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2    0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3    0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4    0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
..        ...       ...       ...       ...       ...       ...       ...   
767 -0.107582 -0.418263 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   
768 -0.161175  0.088496  0.285390  0.281069 -0.370130  0.043410  0.092318   
769 -0.594609  0.159877  0.091873  0.140964  0.227406 -0.017389  0.016030   
770 -0.177225 -0.222918 -1.245505  0.678360  0.525059  0.002920 -0.003333   
771 -0.125231 -0.057041  0.073082  0.633977 -0.310685  0.033590  0.015250   

     Class  
0        0  
1        1  
2        0  
3        0  
4        0  
..     ...  
767      0  
768      0  
769      0  
770      0  
771      0  

[772 rows x 31 columns]

In [28]:
#smote
X = df2.drop('Class', axis=1)
y = df2['Class']
oversample = SMOTE()
#rus = RandomUnderSampler(random_state=42)
#X_rs, y_rs = rus.fit_resample(X,y)
X_smo, y_smo = oversample.fit_resample(X, y)

In [29]:
df_smo = pd.DataFrame(np.hstack((X_smo,y_smo[:, None])), columns=df2.columns)
df_smo.Class = df_smo.Class.astype(int)
df_smo.head()

scaled_amount  scaled_time        V1        V2        V3        V4  \
0       2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1      -0.282095    -0.923077  1.191857  0.266151  0.166480  0.448154   
2       7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
3       2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
4       1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   

         V5        V6        V7        V8  ...       V20       V21       V22  \
0 -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307  0.277838   
1  0.060018 -0.082361 -0.078803  0.085102  ... -0.069083 -0.225775 -0.638672   
2 -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998  0.771679   
3 -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300  0.005274   
4 -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431  0.798278   

        V23       V24       V25       V26       V27       V28  Class  
0 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053      0  
1  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724      1  
2  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752      0  
3 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458      0  
4 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153      0  

[5 rows x 31 columns]

In [30]:
print('Percentage of No Frauds: {}%'.format(round(df_smo.Class.value_counts()[0]/len(df_smo) * 100.0,2)))
print('Percentage of Frauds: {}%'.format(round(df_smo.Class.value_counts()[1]/len(df_smo) * 100.0,2)))

Percentage of No Frauds: 50.0%
Percentage of Frauds: 50.0%



SMOTE + ENN Sampling

In [31]:
df3 = data
df3

scaled_amount  scaled_time        V1        V2        V3        V4  \
0         2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1        -0.282095    -0.923077  1.191857  0.266151  0.166480  0.448154   
2         7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
3         2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
4         1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   
..             ...          ...       ...       ...       ...       ...   
767       0.202967     0.959083 -0.572263  0.731748  1.541254  0.150506   
768       1.278462     0.972177 -1.296845 -0.511605  2.404726 -0.310762   
769      -0.215684     0.972177  1.214170  0.210481  0.484651  0.479768   
770      -0.086899     0.975450  1.267030 -0.071114  0.037680  0.512683   
771      -0.058034     0.978723  1.153758  0.132273  0.382969  1.405063   

           V5        V6        V7        V8  ...       V20       V21  \
0   -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307   
1    0.060018 -0.082361 -0.078803  0.085102  ... -0.069083 -0.225775   
2   -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998   
3   -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300   
4   -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431   
..        ...       ...       ...       ...  ...       ...       ...   
767  1.108974  0.372152  1.084879 -0.146329  ...  0.056544 -0.143508   
768 -0.319551 -0.542842 -0.173310  0.260423  ... -0.349668 -0.071270   
769 -0.261955 -0.527039  0.021782 -0.106888  ... -0.071540 -0.224292   
770  0.242392  0.705212 -0.226582  0.109483  ... -0.066990 -0.164468   
771 -0.224287 -0.197295  0.020653  0.029260  ... -0.230877 -0.107809   

          V22       V23       V24       V25       V26       V27       V28  \
0    0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1   -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2    0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3    0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4    0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
..        ...       ...       ...       ...       ...       ...       ...   
767 -0.107582 -0.418263 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   
768 -0.161175  0.088496  0.285390  0.281069 -0.370130  0.043410  0.092318   
769 -0.594609  0.159877  0.091873  0.140964  0.227406 -0.017389  0.016030   
770 -0.177225 -0.222918 -1.245505  0.678360  0.525059  0.002920 -0.003333   
771 -0.125231 -0.057041  0.073082  0.633977 -0.310685  0.033590  0.015250   

     Class  
0        0  
1        1  
2        0  
3        0  
4        0  
..     ...  
767      0  
768      0  
769      0  
770      0  
771      0  

[772 rows x 31 columns]

In [32]:
#ENN sampling
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.pipeline import Pipeline
from imblearn.combine import SMOTEENN
from imblearn.under_sampling import EditedNearestNeighbours
from sklearn.model_selection import cross_validate

In [33]:
X = df3.drop('Class', axis = 1)
y = df3['Class']
resample=SMOTEENN(enn=EditedNearestNeighbours(sampling_strategy='all'))
Xsmee , ysmee = resample.fit_resample(X,y)

In [34]:
df_smee = pd.DataFrame(np.hstack((Xsmee,ysmee[:, None])), columns=df3.columns)
df_smee.Class = df_smee.Class.astype(int)
df_smee.head()

scaled_amount  scaled_time        V1        V2        V3        V4  \
0       2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1       7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
2       2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
3       1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   
4      -0.262313    -0.916530 -0.425966  0.960523  1.141109 -0.168252   

         V5        V6        V7        V8  ...       V20       V21       V22  \
0 -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307  0.277838   
1 -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998  0.771679   
2 -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300  0.005274   
3 -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431  0.798278   
4  0.420987 -0.029728  0.476201  0.260314  ...  0.084968 -0.208254 -0.559825   

        V23       V24       V25       V26       V27       V28  Class  
0 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053      0  
1  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752      0  
2 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458      0  
3 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153      0  
4 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080      0  

[5 rows x 31 columns]

In [35]:
print('Percentage of No Frauds: {}%'.format(round(df_smee.Class.value_counts()[0]/len(df_smee) * 100.0,2)))
print('Percentage of Frauds: {}%'.format(round(df_smee.Class.value_counts()[1]/len(df_smee) * 100.0,2)))

Percentage of No Frauds: 48.2%
Percentage of Frauds: 51.8%


Stratified Sampling

In [36]:
#stratified Sampling
df4 = data
df4

scaled_amount  scaled_time        V1        V2        V3        V4  \
0         2.683791    -0.923077 -1.359807 -0.072781  2.536347  1.378155   
1        -0.282095    -0.923077  1.191857  0.266151  0.166480  0.448154   
2         7.307126    -0.919804 -1.358354 -1.340163  1.773209  0.379780   
3         2.156540    -0.919804 -0.966272 -0.185226  1.792993 -0.863291   
4         1.076403    -0.916530 -1.158233  0.877737  1.548718  0.403034   
..             ...          ...       ...       ...       ...       ...   
767       0.202967     0.959083 -0.572263  0.731748  1.541254  0.150506   
768       1.278462     0.972177 -1.296845 -0.511605  2.404726 -0.310762   
769      -0.215684     0.972177  1.214170  0.210481  0.484651  0.479768   
770      -0.086899     0.975450  1.267030 -0.071114  0.037680  0.512683   
771      -0.058034     0.978723  1.153758  0.132273  0.382969  1.405063   

           V5        V6        V7        V8  ...       V20       V21  \
0   -0.338321  0.462388  0.239599  0.098698  ...  0.251412 -0.018307   
1    0.060018 -0.082361 -0.078803  0.085102  ... -0.069083 -0.225775   
2   -0.503198  1.800499  0.791461  0.247676  ...  0.524980  0.247998   
3   -0.010309  1.247203  0.237609  0.377436  ... -0.208038 -0.108300   
4   -0.407193  0.095921  0.592941 -0.270533  ...  0.408542 -0.009431   
..        ...       ...       ...       ...  ...       ...       ...   
767  1.108974  0.372152  1.084879 -0.146329  ...  0.056544 -0.143508   
768 -0.319551 -0.542842 -0.173310  0.260423  ... -0.349668 -0.071270   
769 -0.261955 -0.527039  0.021782 -0.106888  ... -0.071540 -0.224292   
770  0.242392  0.705212 -0.226582  0.109483  ... -0.066990 -0.164468   
771 -0.224287 -0.197295  0.020653  0.029260  ... -0.230877 -0.107809   

          V22       V23       V24       V25       V26       V27       V28  \
0    0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1   -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2    0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3    0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4    0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
..        ...       ...       ...       ...       ...       ...       ...   
767 -0.107582 -0.418263 -0.731029  0.877525 -0.364150 -0.177509 -0.256545   
768 -0.161175  0.088496  0.285390  0.281069 -0.370130  0.043410  0.092318   
769 -0.594609  0.159877  0.091873  0.140964  0.227406 -0.017389  0.016030   
770 -0.177225 -0.222918 -1.245505  0.678360  0.525059  0.002920 -0.003333   
771 -0.125231 -0.057041  0.073082  0.633977 -0.310685  0.033590  0.015250   

     Class  
0        0  
1        1  
2        0  
3        0  
4        0  
..     ...  
767      0  
768      0  
769      0  
770      0  
771      0  

[772 rows x 31 columns]

In [37]:
# we create a function to check if the data is proportional based on a pre-defined variable.
def counts(df, var):
    c = df[var].value_counts(dropna=False)
    p = df[var].value_counts(dropna=False, normalize=True)
    return pd.concat([c,p], axis=1, keys=['counts', '%'])

In [38]:
from sklearn.model_selection import StratifiedShuffleSplit as ss

In [39]:
#using the sklearn function to obtain the sample, 
# based on 10% of the stratified elements in the default variable.

sp = ss(test_size=0.1)
for x, y in sp.split(df4, df4['Class']):
    dfx = df4.iloc[x]
    dfy = df4.iloc[y]

#dfx.shape, dfy.shape

print('sample:',dfy.shape[0],'\n')
display(counts(dfy, 'Class'))

sample: 78 



counts         %
0      77  0.987179
1       1  0.012821

In [40]:
# function for extracting stratified samples with useful parameters

def sample_stratified(dataframe, size, variable):
    sp = ss(test_size=size)
    for _, y in sp.split(dataframe, dataframe[variable]):
        dataframey = dataframe.iloc[y]
    return dataframey

In [41]:
df_sample_stratified = sample_stratified(df4, 0.1, 'Class')

display(df_sample_stratified.head(3))
print('\n')

display(df_sample_stratified.shape)
print('\n')

display(counts(df_sample_stratified,'Class'))
print('\n')

scaled_amount  scaled_time        V1        V2        V3        V4  \
271      -0.300262    -0.294599 -0.351287  1.183754  0.530794  0.543891   
362      -0.127473    -0.052373 -2.564961  2.470985  2.649417 -1.564256   
681       0.955491     0.756137 -0.632438 -0.363984  2.354709  0.301993   

           V5        V6        V7        V8  ...       V20       V21  \
271  0.798441 -0.804472  1.563170 -0.935794  ...  0.596401 -0.121755   
362  1.794297 -0.614742  4.185906 -3.855359  ...  3.564138 -1.672706   
681 -0.212088  0.982642 -0.371442  0.222824  ... -0.047421 -0.354384   

          V22       V23       V24       V25       V26       V27       V28  \
271  0.668124 -0.057380  0.301398 -0.730946 -0.538616 -0.037212 -0.479956   
362 -0.463149 -0.532466  0.306494  0.226844 -0.365416 -0.936735 -2.733887   
681 -0.531434 -0.165897 -0.904714  0.159312 -0.187616 -0.048369 -0.136024   

     Class  
271      0  
362      0  
681      0  

[3 rows x 31 columns]

(78, 31)

counts         %
0      77  0.987179
1       1  0.012821

In [42]:
def sample_stratified(dataframe, size, variable):
    p = (size / len(dataframe))
    sp = ss(test_size=p)
    for _, y in sp.split(dataframe, dataframe[variable]):
        dataframey = dataframe.iloc[y]
    return dataframey

In [43]:
df_sample_stratified = sample_stratified(df_smo, 100, 'Class')

display(df_sample_stratified.head(3))
print('\n')

display(df_sample_stratified.shape)
print('\n')

display(counts(df_sample_stratified,'Class'))
print('\n')

scaled_amount  scaled_time        V1        V2        V3        V4  \
804       -0.308055     0.374833 -0.406263  0.386742  1.397238  0.304689   
412       -0.087909     0.049100  1.145614  0.004148  0.772422  0.916543   
1520      -0.284984    -0.788942  1.012371  0.277330  0.299799  0.425677   

            V5        V6        V7        V8  ...       V20       V21  \
804   0.646586 -0.941462  0.689640 -0.167644  ... -0.109074 -0.018254   
412  -0.568699 -0.281603 -0.161752 -0.025480  ... -0.098173 -0.046280   
1520  0.136756 -0.151672  0.002291  0.064639  ... -0.073249 -0.201042   

           V22       V23       V24       V25       V26       V27       V28  \
804   0.015073 -0.165685  0.374398  0.349244 -0.322726 -0.085063 -0.088318   
412   0.174420 -0.014831  0.481153  0.460998  0.417277  0.007883  0.014894   
1520 -0.560775  0.070872 -0.278444  0.186131  0.076810 -0.017066  0.002830   

      Class  
804       1  
412       0  
1520      1  

[3 rows x 31 columns]

(100, 31)

counts    %
1      50  0.5
0      50  0.5


Systematic Sampling
Taking balanced data from smote

In [44]:
size = 100
length = len(df_smo)
interval = length // size
print('interval:',interval)

interval: 15


In [45]:

import random as rd
#define the seed of random. 'None' paramn will allow random change every execution.
rd.seed(None) #or rd.seed(1) to change the number every execution this part of the code.

first = rd.randint(0, interval)
print('first element:',first)

first element: 7


In [46]:
indexes = np.arange(first, length, step = interval)
indexes

array([   7,   22,   37,   52,   67,   82,   97,  112,  127,  142,  157,
        172,  187,  202,  217,  232,  247,  262,  277,  292,  307,  322,
        337,  352,  367,  382,  397,  412,  427,  442,  457,  472,  487,
        502,  517,  532,  547,  562,  577,  592,  607,  622,  637,  652,
        667,  682,  697,  712,  727,  742,  757,  772,  787,  802,  817,
        832,  847,  862,  877,  892,  907,  922,  937,  952,  967,  982,
        997, 1012, 1027, 1042, 1057, 1072, 1087, 1102, 1117, 1132, 1147,
       1162, 1177, 1192, 1207, 1222, 1237, 1252, 1267, 1282, 1297, 1312,
       1327, 1342, 1357, 1372, 1387, 1402, 1417, 1432, 1447, 1462, 1477,
       1492, 1507, 1522])

In [47]:
df_sample_systematic = df_smo.iloc[indexes]
display(df_sample_systematic.head(3))
print('\n')

display(df_sample_systematic.shape)
print('\n')

scaled_amount  scaled_time        V1        V2        V3        V4  \
7        0.487182    -0.900164 -0.644269  1.417964  1.074380 -0.492199   
22      -0.290371    -0.864157  1.166616  0.502120 -0.067300  2.261569   
37      -0.300061    -0.834697 -1.452187  1.765124  0.611669  1.176825   

          V5        V6        V7        V8  ...       V20       V21       V22  \
7   0.948934  0.428118  1.120631 -3.807864  ... -0.156742  1.943465 -1.015455   
22  0.428804  0.089474  0.241147  0.138082  ... -0.307169  0.018702 -0.061972   
37 -0.445980  0.246826 -0.257566  1.092472  ...  0.008713  0.082280  0.325782   

         V23       V24       V25       V26       V27       V28  Class  
7   0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339      0  
22 -0.103855 -0.370415  0.603200  0.108556 -0.040521 -0.011418      0  
37 -0.069107  0.020962 -0.044668 -0.243441  0.149180  0.120557      0  

[3 rows x 31 columns]

(102, 31)

In [48]:
def sample_systematic(dataframe, sample_size, state = None):
    length = len(df_smo)
    interval = length // sample_size
    rd.seed(state)
    first = rd.randint(0, interval)
    indexes = np.arange(first, length, step = interval)
    return dataframe.iloc[indexes]

In [49]:
df_sample_systematic = sample_systematic(df_smo, 100)

display(df_sample_systematic.head(3))
print('')

display(df_sample_systematic.shape)
print('')

scaled_amount  scaled_time        V1        V2        V3        V4  \
12       2.116169    -0.890344  1.249999 -1.221637  0.383930 -1.234899   
27      -0.013423    -0.847791  1.322707 -0.174041  0.434555  0.576038   
42      -0.037646    -0.815057 -0.607877  1.031345  1.740450  1.232106   

          V5        V6        V7        V8  ...       V20       V21       V22  \
12 -1.485419 -0.753230 -0.689405 -0.227487  ... -0.102756 -0.231809 -0.483285   
27 -0.836758 -0.831083 -0.264905 -0.220982  ... -0.522951 -0.284376 -0.323357   
42  0.418592  0.119168  0.850893 -0.176267  ...  0.254325 -0.087329  0.258315   

         V23       V24       V25       V26       V27       V28  Class  
12  0.084668  0.392831  0.161135 -0.354990  0.026416  0.042422      0  
27 -0.037710  0.347151  0.559639 -0.280158  0.042335  0.028822      0  
42 -0.264775  0.118282  0.173508 -0.217041  0.094312 -0.033041      0  

[3 rows x 31 columns]

(101, 31)

In [51]:
print('Percentage of No Frauds: {}%'.format(round(df_sample_systematic.Class.value_counts()[0]/len(df_sample_systematic) * 100.0,2)))
print('Percentage of Frauds: {}%'.format(round(df_sample_systematic.Class.value_counts()[1]/len(df_sample_systematic) * 100.0,2)))

Percentage of No Frauds: 49.5%
Percentage of Frauds: 50.5%


Applying Classifiers Models on these sampled datasets
I'll be using 5 different classifiers to classify the transactions as Fraudulant or Non-Fraudulant in the randomly undersampled dataset. My aim is to compare the performance of the Naive predictor with that of the classifiers that I choose. The following are the classifiers that I'll be using:-

Logistic Regression
Support Vector Classifier
Decision Tree
KNearest Classifier
SGD Classifier

In [52]:
array_random_sampling  = []
array_smote = []
array_smee = []
array_systematic = []
array_stratified = []

Models on Random Under Sampling

In [53]:
# ML models on Random Under Sampling
# Undersampling before cross validation(prove to overfit)
X = df_rs.drop('Class', axis=1)
y = df_rs['Class'] 

In [54]:
from sklearn.linear_model import SGDClassifier
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [55]:
#Convert to arrays to feed to the classifications algorithms
classifiers = {
    "Logistic Regression": LogisticRegression(),
    "Support Vector Classifier": SVC(),
    "Decision Tree": DecisionTreeClassifier(),
    "KNearest": KNeighborsClassifier(),
    "SGD Classifier": SGDClassifier()
    
    
}

In [56]:
print('Cross-Validation Scores:-')
for key, classifier in classifiers.items():
    classifier.fit(X_train, y_train)
    cv_score = cross_val_score(classifier, X_train, y_train, cv=5)
    print('{}: {}'.format(key,round(cv_score.mean()*100.0, 2)))
    array_random_sampling.append(round(cv_score.mean()*100.0, 2))

Cross-Validation Scores:-
Logistic Regression: 80.0
Support Vector Classifier: 63.33
Decision Tree: 66.67
KNearest: 36.67
SGD Classifier: 80.0


In [57]:
array_random_sampling

[80.0, 63.33, 66.67, 36.67, 80.0]

Models on Smote Over Sampling

In [58]:
#SMOTE models
X = df_smo.drop('Class', axis=1)
y = df_smo['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [59]:
print('Cross-Validation Scores:-')
for key, classifier in classifiers.items():
    classifier.fit(X_train, y_train)
    cv_score = cross_val_score(classifier, X_train, y_train, cv=5)
    print('{}: {}'.format(key,round(cv_score.mean()*100.0, 2)))
    array_smote.append(round(cv_score.mean()*100.0, 2))

Cross-Validation Scores:-
Logistic Regression: 92.54
Support Vector Classifier: 97.54
Decision Tree: 97.54
KNearest: 95.08
SGD Classifier: 88.85


Models on SMOTE + ENN Sampling

In [60]:
#ML on SMEE sampling
#SMOTE models
X = df_smee.drop('Class', axis=1)
y = df_smee['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [61]:
print('Cross-Validation Scores:-')
for key, classifier in classifiers.items():
    classifier.fit(X_train, y_train)
    cv_score = cross_val_score(classifier, X_train, y_train, cv=5)
    print('{}: {}'.format(key,round(cv_score.mean()*100.0, 2)))
    array_smee.append(round(cv_score.mean()*100.0, 2))

Cross-Validation Scores:-
Logistic Regression: 93.8
Support Vector Classifier: 98.22
Decision Tree: 98.13
KNearest: 96.86
SGD Classifier: 94.39



Models on Sample Stratified Sampling

In [62]:
#ML on Sample Stratified
#sample stratifies
X = df_sample_stratified.drop('Class', axis=1)
y = df_sample_stratified['Class']

In [63]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [64]:
print('Cross-Validation Scores:-')
for key, classifier in classifiers.items():
    classifier.fit(X_train, y_train)
    cv_score = cross_val_score(classifier, X_train, y_train, cv=5)
    print('{}: {}'.format(key,round(cv_score.mean()*100.0, 2)))
    array_stratified.append(round(cv_score.mean()*100.0, 2))

Cross-Validation Scores:-
Logistic Regression: 88.75
Support Vector Classifier: 91.25
Decision Tree: 95.0
KNearest: 72.5
SGD Classifier: 92.5


Models on Sample Systematic Sampling

In [65]:
#ML on Sample Systematic
#sample systematic
X = df_sample_systematic.drop('Class', axis=1)
y = df_sample_systematic['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [66]:
print('Cross-Validation Scores:-')
for key, classifier in classifiers.items():
    classifier.fit(X_train, y_train)
    cv_score = cross_val_score(classifier, X_train, y_train, cv=5)
    print('{}: {}'.format(key,round(cv_score.mean()*100.0, 2)))
    array_systematic.append(round(cv_score.mean()*100.0, 2))

Cross-Validation Scores:-
Logistic Regression: 81.25
Support Vector Classifier: 88.75
Decision Tree: 88.75
KNearest: 71.25
SGD Classifier: 91.25


In [67]:
keys = ['Random Sampling','Smote','SMEE Oversampling','Sample Stratified','Sample Systematic']

Making a table for defined view of CV Score of Models on different samplings

In [69]:
from tabulate import tabulate
print(tabulate({'Classifiers Models':['Logistic Regression','Support Vector Classifier','Desicion Tree','KNearest','SGD Classifier'],
    'Random Sampling': array_random_sampling, 
                'Smote': array_smote, 
                'Smee Oversampling': array_smee,
               'Sample Stratified': array_stratified,
               'Sample Systematic': array_systematic}, 
                headers='keys', 
                tablefmt='fancy_grid'))

╒═══════════════════════════╤═══════════════════╤═════════╤═════════════════════╤═════════════════════╤═════════════════════╕
│ Classifiers Models        │   Random Sampling │   Smote │   Smee Oversampling │   Sample Stratified │   Sample Systematic │
╞═══════════════════════════╪═══════════════════╪═════════╪═════════════════════╪═════════════════════╪═════════════════════╡
│ Logistic Regression       │             80    │   92.54 │               93.8  │               88.75 │               81.25 │
├───────────────────────────┼───────────────────┼─────────┼─────────────────────┼─────────────────────┼─────────────────────┤
│ Support Vector Classifier │             63.33 │   97.54 │               98.22 │               91.25 │               88.75 │
├───────────────────────────┼───────────────────┼─────────┼─────────────────────┼─────────────────────┼─────────────────────┤
│ Desicion Tree             │             66.67 │   97.54 │               98.13 │               95    │               